File: Joueur.ts/src/joueur/base-ai.ts
import { IBasePlayer } from "@cadre/ts-utils/cadre";
import * as chalk from "chalk";
import { BaseGame } from "./base-game";

/**
 * A helper function to set "private" properties of the AI classes
 * @param ai the AI instance to set settings inside of
 * @param aiSettings unparsed key/value strings to set
 */
export function setAISettings(ai: BaseAI, aiSettings: string): void {
    if (aiSettings) {
        const settings = aiSettings.split("&");
        for (const setting of settings) {
            const kv = setting.split("=");
             // cast to any to use private field
            (ai as any).settings[kv[0]] = kv[1];
        }
    }
}

/**
 * The base interactions all AIs require to work on any game
 */
export class BaseAI {
    /** The base game this AI is player */
    public readonly game!: BaseGame;

    /** The base player this AI controls */
    public readonly player!: IBasePlayer;

    // will be set by the setAISettings function above
    private settings: { [key: string]: string | undefined } = {};

    /**
     * Creates an instance of the AI. Competitors should never invoke this.
     * @hidden
     * @param game The game this AI is playing.
     */
    public constructor(game: BaseGame) {
        this.game = game;
    }

    /**
     * A static method to get the name of the player for this AI before
     * gameplay starts.
     * @returns A string of their name.
     */
    public getName(): string {
         // Intended to be overridden by the AI class
        return "TypeScript Player";
    }

    /**
     * Called when the game starts, after the first delta is sent. Kind of a
     * lazy constructor as at construction time we don't know our player.
     */
    public async start(): Promise<void> {
        // Intended to be overridden by the AI class
    }

    /**
     * Invoked after a delta update happens (for any player in the game).
     */
    public async gameUpdated(): Promise<void> {
        // Intended to be overridden by the AI class
    }

    /**
     * Invoked after this AI sends some run that is invalid
     * @param message The human readable error message about why it did
     * something invalid.
     */
    public async invalid(message: string): Promise<void> {
        // tslint:disable-next-line:no-console
        console.warn(chalk.yellow(`Invalid: ${message}`));
    }

    /**
     * Invoked when the game is over. Think of this as a deconstructor.
     * @param won If this AI's player won the game or not.
     * @param reason The human readable reason why they won or lost the game.
     */
    public async ended(won: boolean, reason: string): Promise<void> {
        // Intended to be overridden by the AI class
    }

    /**
     * Gets an AI setting passed to the program via the `--aiSettings` flag.
     * If the flag was set it will be returned as a string value, undefined
     * otherwise.
     *
     * @param key - The key of the setting you wish to get the value
     * for.
     * @returns A string representing the value set via
     * command line, or undefined if the key was not set.
     */
    public getSetting(key: string): string | undefined {
        return this.settings[key];
    }
}

File: Joueur.go/base/base_ai.go
package base

import (
	"github.com/fatih/color"
)

// AISettings is kind of hacky; basically exposing a global for all AI structs,
// however only 1 AI should ever be constructed per game so should be an ok assumption
var AISettings map[string]([]string)

// AI is the base interface all game AIs should implement to correctly
// interface and play their game.
type AI interface {
	Start()
	Ended(bool, string)
	GameUpdated()
	Invalid(string)
}

// AIImpl is the implimentation struct for the AI interface
type AIImpl struct {
	game   Game
	player Player
}

// Game returns the game this [base] AI is in
func (ai *AIImpl) Game() Game {
	return ai.game
}

// Player returns the game this [base] AI is in
func (ai *AIImpl) Player() Player {
	return ai.player
}

// Start is called once the game starts and your AI knows its Player and
// Game. You can initialize your AI here.
func (ai AIImpl) Start() {
	// pass
}

// Ended is called when the game ends.
// You can clean up your data and dump files here if need be.
func (ai AIImpl) Ended(won bool, reason string) {
	// pass
}

// GameUpdated is called every time the game's state updates,
// so if you are tracking anything you can update it here.
func (ai AIImpl) GameUpdated() {
	// pass
}

// Invalid is automatically called after this AI sends some run that is
// has arguments that are invalidated for some reason.
func (ai AIImpl) Invalid(message string) {
	color.Yellow("Invalid: " + message)
}

// GetSetting gets an AI setting passed to the program via the --aiSettings
// flag. If the flag was set it will be returned as a string value, undefined otherwise.
func (ai AIImpl) GetSetting(key string) ([]string, bool) {
	setting, ok := AISettings[key]
	return setting, ok
}

// InjectIntoAI injects into an AI its "private" player and game
func InjectIntoAI(ai *AIImpl, game Game, player Player) {
	ai.game = game
	ai.player = player
}
